src/xma/src/xmaapi/xmakernel.cpp

```diff
@@ -240,39 +195,57 @@ xma_kernel_session_create(XmaKernelProperties *props)
     priv1->kernel_execbos.reserve(num_execbo);
     priv1->num_execbo_allocated = num_execbo;
     if (xma_core::create_session_execbo(priv1, num_execbo, XMA_KERNEL_MOD) != XMA_SUCCESS) {
-        //Release singleton lock
-        g_xma_singleton->locked = false;
         free(session->base.plugin_data);
         free(session);
         delete priv1;
         return nullptr;
     }
 
-    rc = session->kernel_plugin->init(session);
-    if (rc) {
-        xma_logmsg(XMA_ERROR_LOG, XMA_KERNEL_MOD,
-                   "Initalization of kernel plugin failed. Return code %d\n",
-                   rc);
-        //Release singleton lock
-        g_xma_singleton->locked = false;
-        free(session->base.plugin_data);
-        free(session);
-        delete priv1;
-        return nullptr;
+    //Obtain lock only for a) singleton changes & b) kernel_info changes
+    std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);
+    //Singleton lock acquired
+
+    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
+        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
+            xma_logmsg(XMA_ERROR_LOG, XMA_KERNEL_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
+            free(session->base.plugin_data);
+            free(session);
+            delete priv1;
```

Comment:
I am worried about random frees and deletes.  This does not look right.  Several other places too.

Reply:
Based on comments that this has many change. I will setup separate review of couple of file at a time.